patroni/ha.py

```diff
@@ -1402,3 +1415,23 @@ def get_remote_member(self, member=None):
 
     def get_remote_master(self):
         return self.get_remote_member()
+
+    def call_pre_promote(self, config):
+        """
+        Runs a fencing script after the leader lock is acquired but before the replica is promoted.
+        If the script exits with a non-zero code, promotion does not happen and the leader key is removed from DCS.
+        """
+        cmd = config.get('pre_promote')
+        if cmd:
+            self._pre_promote_subprocess = CancellableSubprocess()
```

Comment:
The idea of CancallableSubprocess is that there is only one instance of it, which is also used by [AsyncExecutor](https://github.com/zalando/patroni/blob/master/patroni/ha.py#L77), what makes it possible to easily cancel it.

Reply:
Is it because having a single process simplifies the `AsyncExecutor` (is > 1 critical simultaneous task actually ever needed ?) or are there OS-level reasons for that ?